### ISSUE TITLE
'(string)null!' in parameter default value gives unexpected warning

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#47821]
Follow-up to #32844

**Version Used**: e6475bc8f7aa60ffd9ca1de2f4096166117f16aa

**Steps to Reproduce**:

```cs
#nullable enable

public class C
{
    public void M(string s = (string)null!) // unexpected warning
    {
    }
}
```

**Expected Behavior**: No warnings

**Actual Behavior**: Warning on the default value

This will be a lot easier to get right if/when parameter default values are part of the bound tree #47352 (we can visit the bound default value simply as an expression, and then check assignment of the "lowered" constant value to the parameter).

**Comment by RikkiGibson** [comment]
In general, suppressions are being ignored when they are not on the top level. We can fall back to checking default value syntax to "hack around" this. But I think edge cases with suppressions in parameter default values are not very high priority to us.

**Comment by Youssef1313** [comment]
This could be tweaked but I think the warning is fine here. Suppression has higher precedence than cast, so the suppression is done first, then there is an attempt to cast from null to non-nullable type.

To be more clear about what I mean:

```csharp
public void M(string s = (string)(null!)) // Warning is expected. suppression then casting.
public void M(string s = ((string)null)!) // Warning is unexpected. casting then suppression.
```

**Comment by RikkiGibson** [comment]
The expression `null!` has not-null state, so no warning is expected when casting it to non-nullable reference type.

**Comment by RikkiGibson** [comment]
Fixed by #49009
